src/include/miopen/graphapi/opgraph.h

```diff
@@ -0,0 +1,223 @@
+
+    friend class OpGraphBuilder;
+    friend class OpGraph;
+
+protected:
+    virtual const std::string& signName() const = 0;
+
+    virtual std::vector<Tensor*> getInTensors() const = 0;
+
+    virtual std::vector<Tensor*> getOutTensors() const = 0;
+
+    const auto& iterateInEdges() const { return mInEdges; }
+
+    const auto& iterateOutEdges() const { return mOutEdges; }
+
+    bool hasInEdge(OpNode* src, Tensor* tens_ptr) const
+    {
+        Edge e{src, tens_ptr};
+        return internal::contains(mInEdges, e);
+    }
+
+    bool hasOutEdge(OpNode* dst, Tensor* tens_ptr) const
+    {
+        Edge e{dst, tens_ptr};
+        return internal::contains(mOutEdges, e);
+    }
+
+    void addOutEdge(OpNode* dst, Tensor* tens_ptr)
+    {
+        assert(dst);
+        if(!hasOutEdge(dst, tens_ptr))
+        {
+            mOutEdges.emplace_back(dst, tens_ptr);
+        }
+    }
+
+    void addInEdge(OpNode* src, Tensor* tens_ptr)
+    {
+        assert(src);
+        if(!hasInEdge(src, tens_ptr))
+        {
+            mInEdges.emplace_back(src, tens_ptr);
+        }
+    }
+};
+
+using OpEdge = OpNode::Edge;
+
+class SourceOpNode : public OpNode
+{
+protected:
+    std::vector<Tensor*> mOutTensors;
+    friend class OpGraph;
+
+    const std::string& signName() const final
+    {
+        static const std::string s = "INTERNAL::SRC";
+        return s;
+    }
+
+    std::vector<Tensor*> getInTensors() const final { return {}; }
+
+    std::vector<Tensor*> getOutTensors() const final { return mOutTensors; }
+
+    bool hasOutTensor(Tensor* tensor) const { return internal::contains(mOutTensors, tensor); }
+
+    void addOutTensor(Tensor* tens_ptr)
+    {
+        assert(!hasOutTensor(tens_ptr));
+        mOutTensors.emplace_back(tens_ptr);
+    }
+};
+
+class SinkOpNode : public OpNode
+{
+protected:
+    std::vector<Tensor*> mInTensors;
+    friend class OpGraph;
+
+    const std::string& signName() const final
+    {
+        static const std::string s = "INTERNAL::SINK";
+        return s;
+    }
+
+    std::vector<Tensor*> getInTensors() const final { return mInTensors; }
+
+    std::vector<Tensor*> getOutTensors() const final { return {}; }
+
+    bool hasInTensor(Tensor* tensor) const { return internal::contains(mInTensors, tensor); }
+
+    void addInTensor(Tensor* tens_ptr)
+    {
+        assert(!hasInTensor(tens_ptr));
+        mInTensors.emplace_back(tens_ptr);
+    }
+};
+
+class OpGraph
+{
+    SourceOpNode mSrcNode{};
+    SinkOpNode mSinkNode{};
+    std::vector<OpNode*> mNodes{};
+
+public:
+    bool hasNode(OpNode* n) const { return internal::contains(mNodes, n); }
+
+    bool hasEdge(OpNode* src, Tensor* tens_ptr, OpNode* dst) const
+    {
+        assert(src);
+        assert(dst);
+        return src->hasOutEdge(dst, tens_ptr) && dst->hasInEdge(src, tens_ptr);
+    }
+
+private:
+    friend class OpGraphBuilder;
+
+    void addNodes(std::vector<OpNode*>&& nodes) { mNodes = std::move(nodes); }
```

Comment:
It looks like `setNodes`, since it replaces the previous state.

Reply:
yeah, I need a better name. Perhaps `moveNodes` because `setNodes` doesn't quite cut it. I hate setters in general.